### PR TITLE
[WIP] Warmup task JSUserRegistrationService App

### PR DESCRIPTION
**Pull request by karim-elngr**
### Initial implementation of the SkyPal Warmup task.

**_Motive_**: Getting familiar with the Jumia tech stack. 
### _**How to**_: 

1. `sudo docker-compose up` (mongodb 3.2)
2. `git clone https://github.com/motamira/spring-boot-warmup.git`
3. `cd spring-boot-warmup` 
4. `git checkout karim`
5. `cd skypalwarmup`
## To run
6. `mvn clean install -Pcheckstyle -pl JSUserRegistrationService -am`
7. `sudo java -jar ./JSUserRegistrationService/target/JSUserRegistrationService-1.0-SNAPSHOT.jar`
## To test
8. or `mvn clean test -Pcheckstyle -pl JSUserRegistrationService -am`

### CODE REVIEW

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/UserController.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+@RequestMapping("/api/users")
+@AllArgsConstructor
+public class UserController {
+
+    private static final Log logger = LogFactory.getLog(UserController.class);
+
+    /**
+     * The result returned to the user after calling the registration service, includes a status code and a message.
+     */
+    public enum Result {
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/UserController.java]: Please move this enum to a new file (e.g.  enums/ResultEnum.java)

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/AccountInformationDTO.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class AccountInformationDTO {
+    @NotNull
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/AccountInformationDTO.java]: Please add a new line here after class opening and before @annotation (same comment in most of other classes..please check the code formatter)

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/ContactsDTO.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class ContactsDTO {
+    @Nullable
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/ContactsDTO.java]: new line

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/PersonalDetailsDTO.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class PersonalDetailsDTO {
+    @NotNull
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/PersonalDetailsDTO.java]: new line

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class UserInputDTO {
+    @NotNull
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java]: new line

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/AccountInformation.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class AccountInformation {
+    @Indexed(unique = true)
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/AccountInformation.java]: new line

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/Contacts.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class Contacts {
+    private String cellPhone;
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/Contacts.java]: add new line

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/User.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Document
+public class User {
+    @Id
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/User.java]: add new line

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.NoArgsConstructor;
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class UserInputDTO {
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java]: is "Input" word necessary? can be simply named "UserDTO" what do you think?

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/service/UserService.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package application.service;
+
+import application.dto.UserInputDTO;
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+public interface UserService {
+    public Boolean register(UserInputDTO user);
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/service/UserService.java]: Unnecessary access modifier "public" for register method. it is public by default.

**Review by adawood — CHANGES_REQUESTED** [review]

File: skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        userRepository.deleteAll();
+    }
+
+    private void createUserUnderTest() {
+        AccountInformation inTestAccountInformation = new AccountInformation();
+        inTestAccountInformation.setUserName("sho9meUae");
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java]: please add a constant for test strings in the same class or in a separate TestConstants class like `private static final String TEST_USERNAME = "testuser";`

File: skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        inTestAccountInformation.setPassword("somePas$sw1ord");
+
+        Contacts inTestContacts = new Contacts();
+        inTestContacts.setEmail("safgdgom8hgejEfmgailm@ail.com");
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java]: Same comment for string constants in email, phone, firstName, lastName, password, age..etc

**Review by adawood — CHANGES_REQUESTED** [review]

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java (hunk 9 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java]: I was thinking about if we needed to return a UserDTO upon successful registration, the fields would be different in case we wanted to include the user id or other fields we generate and remove the password field.

File: docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,9 @@
+mongo:
```
Comment (motamira) [inline docker-compose.yml:1]: Line 1: Nice initiative in configuring containers even if it is a warmup app :) Well done.

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java (hunk 9 above)
Comment (adawood) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java]: No problem, but In this case it could be simply a "User" entity with the ID and without the password or a masked one :)

File: skypalwarmup/JSUserRegistrationClient/JSUserRegistrationClient.iml [hunk 14]
```diff
@@ -0,0 +1,75 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module org.jetbrains.idea.maven.project.MavenProjectsManager.isMavenModule="true" type="JAVA_MODULE" version="4">
+  <component name="NewModuleRootManager" LANGUAGE_LEVEL="JDK_1_8">
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationClient/JSUserRegistrationClient.iml]: The .iml files don't need to be versioned. You can add these to the .gitignore file. Thanks!

File: skypalwarmup/pom.xml [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+  <packaging>pom</packaging>
+  <version>1.0-SNAPSHOT</version>
+  <modules>
+    <module>JSUserRegistrationService</module>
+    <module>JSUserRegistrationClient</module>
+  </modules>
+
+  <profiles>
+    <profile>
+      <id>checkstyle</id>
```
Comment (motamira) [inline skypalwarmup/pom.xml:19]: Line 18: Interesting. We are using check style in one project only until now. However, I think it gives us several benefits. SkyPal (The service that you will probably develop in afterwards) does not have, atm, checkstyle. It would be a nice addition there :p

Ah, you can check out our current check style configuration here: 

https://github.com/AIGEXP/JS-SkyPro/blob/development/backend/config/checkstyle/jumiaservices_checks.xml

Feel free to give suggestions or any other type of feedback.

**Comment by karim-elngr** [comment]
New updates: 
1. Add JSUserRegistrationClient module
2. Refactor Result Enum
3. Address Styling issues

To run JSUserRegistrationClient

1. mvn clean install -Pcheckstyle -pl JSUserRegistrationClient -am
2. sudo java -jar ./JSUserRegistrationClient/target/JSUserRegistrationClient-1.0-SNAPSHOT.jar

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/RegistrationActionResponses.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
+package application.controller;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+public enum RegistrationActionResponses {
```
Comment (karim-elngr) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/RegistrationActionResponses.java:10]: @adawood This is the new refactored Enum

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/UserController.java (hunk 1 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/UserController.java]: done

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/AccountInformationDTO.java (hunk 2 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/AccountInformationDTO.java]: done

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/ContactsDTO.java (hunk 3 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/ContactsDTO.java]: done

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/PersonalDetailsDTO.java (hunk 4 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/PersonalDetailsDTO.java]: done

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java (hunk 5 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java]: done

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/AccountInformation.java (hunk 6 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/AccountInformation.java]: done

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/Contacts.java (hunk 7 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/Contacts.java]: done

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/User.java (hunk 8 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/User.java]: done

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/UserController.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/UserController.java]: Line 20: Extra blank line here :)

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/UserController.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+@RestController
+@RequestMapping("/api/users")
+@AllArgsConstructor
+public class UserController {
+
+    private static final Log logger = LogFactory.getLog(UserController.class);
+
+    @Autowired
+    UserService userService;
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/UserController.java]: Line 29: Prefer to use constructor injection over field injection and avoid latter when possible. Main advantages is that, this way, you guarantee that your class `UserController` is only instantiated once it has everything it needs. Also, constructor injection makes it easier for you to UT since you can just create mocks and pass them via the constructor. You don't even need Spring at all.

Check this thread for support: https://stackoverflow.com/questions/39890849/what-exactly-is-field-injection-and-how-to-avoid-it

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dao/UserRepository.java [hunk 19]
```diff
@@ -0,0 +1,12 @@
+package application.dao;
+
+import application.entity.User;
+import org.springframework.data.mongodb.repository.MongoRepository;
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/dao/UserRepository.java]: Line 9: Extra blank line here :)

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/AccountInformationDTO.java [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+import javax.validation.constraints.Pattern;
+import javax.validation.constraints.Size;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/AccountInformationDTO.java]: Line 14: Extra blank line here.

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/PersonalDetailsDTO.java [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class PersonalDetailsDTO {
+
+    @NotNull
+    private String firstName;
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/PersonalDetailsDTO.java]: Line 22: What if the name comes empty? Is it valid? :p

To fix this we can use the @NotEmpty instead.

Comment (motamira) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/PersonalDetailsDTO.java]: BTW, same comment for fields that have the @NotNull.

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java (hunk 9 above)
Comment (motamira) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/dto/UserInputDTO.java]: Let me just add that, in fact, IN and OUT DTOs should and would be different ones ;)

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/User.java [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/entity/User.java]: Line 12: Extra blank line here ;)

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/service/UserServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Service
+public class UserServiceImpl implements UserService {
+
+    private static final Log logger = LogFactory.getLog(UserServiceImpl.class);
+
+    @Autowired
+    UserRepository userRepository;
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/service/UserServiceImpl.java]: Line 22: Same comment as the above regarding constructor injection.

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/service/UserServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Override
+    public Boolean register(UserInputDTO userInputDTO) {
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/main/java/application/service/UserServiceImpl.java]: Line 25: This service belongs to our domain logic. Hence, it should already receive a `User` and not a `UserInputDTO`. The reason is that we don't want to couple payload changes (DTO) to our domain. Hence, the `mapFromUserDTO` should be done before the service called. E.g., the `Controller` might call a `Mapper` class that receives a `UserInputDTO` and returns a `User`. The `User` would be then passed to the `register` method.

Alternatively, we could create a `Facade` that would receive a `UserInputDTO`, maps it and, internally, calls the service for user registration. What do you think?

To sum up: Avoid passing DTO to our domain model.

File: skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java]: Line 30: If I'm not mistaken this will load the whole Spring context. If the intention here is to *Unit Test* the controller, and hence it makes sense to mock the `UserService` not the `UserRepository` (Otherwise you are testing more than one component), I think you could use the `@WebMvcTest(UserController.class)` instead.

Check out this link for support: http://www.baeldung.com/spring-boot-testing

File: skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+        userUnderTest.setPersonalDetails(inTestPersonalDetails);
+    }
+
+    private String getUserValueAsJSONString() throws Exception {
+        ObjectMapper om = new ObjectMapper();
+        return om.writeValueAsString(userUnderTest);
+    }
+
+    @Test
+    public void testRegisterCREATED() throws Exception {
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java]: Line 81: For future reference our test naming convention follows something like:

`methodName_whenSomething_returnsSomething` OR `methodName_withSomething_thenSomething`. In this case, we would have:

`register_whenUserDoesNotExist_returnsUserCreated`

File: skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+        this.mockMvc.perform(MockMvcRequestBuilders.post("/api/users").content(getUserValueAsJSONString())
+            .contentType(MediaType.APPLICATION_JSON))
+            .andExpect(MockMvcResultMatchers.status().isCreated())
+            .andExpect(MockMvcResultMatchers.content().string(RegistrationActionResponses.CREATED.statusMessage))
+            .andDo((result) -> Assert.assertTrue(userRepository.findOne(Example.of(userUnderTest)).isPresent()))
+            .andReturn();
+    }
+
+    @Test
+    public void testRegisterBAD_REQUEST() throws Exception {
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java]: Line 91: `register_withInvalidEmail_returnsBadRequest`

File: skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+            .contentType(MediaType.APPLICATION_JSON))
+            .andExpect(MockMvcResultMatchers.status().isBadRequest())
+            .andExpect(MockMvcResultMatchers.content().string(RegistrationActionResponses.BAD_REQUEST.statusMessage))
+            .andReturn();
+        userUnderTest.getContacts().setEmail("not an email");
+        userUnderTest.getContacts().setEmail(oldEmail);
+    }
+
+    @Test
+    public void testRegisterCONFLICT() throws Exception {
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java]: Line 104: `register_whenUserAlreadyExists_returnsConflict`

File: skypalwarmup/JSUserRegistrationClient/src/main/java/application/messaging/configuration/AmqpBindingConfiguration.java [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Slf4j
+@Configuration
+public class AmqpBindingConfiguration {
+
+
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationClient/src/main/java/application/messaging/configuration/AmqpBindingConfiguration.java]: Line 26: Extra blank line here ;)

File: skypalwarmup/JSUserRegistrationClient/src/main/java/application/messaging/dto/PersonalDetailsDTO.java [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class PersonalDetailsDTO {
+
+    @NotNull
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationClient/src/main/java/application/messaging/dto/PersonalDetailsDTO.java]: Line 20: What if this is empty?

File: skypalwarmup/JSUserRegistrationClient/src/main/java/application/service/RestUserRegistrationConsumerImpl.java [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Bean
+    public RestTemplate restTemplate() {
+        return new RestTemplate();
+    }
+
+    public boolean postUserRegistration(UserInputDTO userInputDTO) {
+        try {
+            ResponseEntity<String> response = restTemplate.exchange(RequestEntity
+                .post(new URI("http://localhost:8080/api/users"))
```
Comment (motamira) [inline skypalwarmup/JSUserRegistrationClient/src/main/java/application/service/RestUserRegistrationConsumerImpl.java]: Line 34: This is fine but, of course in a real application this would have to go to the configuration file.

**Review by motamira — CHANGES_REQUESTED** [review]

File: docker-compose.yml (hunk 13 above)
Comment (karim-elngr) [reply docker-compose.yml:1]: I usually am too lazy to setup and configure dependencies (dbs, caches, rabbitmq, etc) on my development machine and always end up reinstalling them for different projects. I am used to running my dependencies using containers, exposing the right ports :D

File: skypalwarmup/pom.xml (hunk 15 above)
Comment (karim-elngr) [reply skypalwarmup/pom.xml:19]: Yes that would be interesting to look into. 
I don't have access to that project, I am getting a 404 after opening that link.

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/UserController.java (hunk 18 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/controller/UserController.java]: Yes, that makes a lot of sense :)

File: skypalwarmup/pom.xml (hunk 15 above)
Comment (motamira) [reply skypalwarmup/pom.xml:19]: Will take care of that

File: skypalwarmup/JSUserRegistrationService/src/main/java/application/service/UserServiceImpl.java (hunk 24 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/main/java/application/service/UserServiceImpl.java]: I think a `Facade` is an overkill in this case, since we only have 1 service, and one type of DTO which maps to a `User`. I like the first approach, and it makes a lot of sense. I was thinking about making domain models for the service layer separate from the Entities which could be very specific to the persistence layer, but for the sake of the warmup, it's not needed.

**Comment by karim-elngr** [comment]
@adawood @motamira All requested changes are now addressed, please review :)

File: skypalwarmup/JSUserRegistrationService/src/test/java/application/TestData.java [hunk 32]
```diff
@@ -0,0 +1,33 @@
+package application;
+
+/**
+ * Copyright (c) 2016-2018, Jumia.
+ */
+public final class TestData {
+
+    public static final class User {
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/test/java/application/TestData.java]: Nice way of organization for accessing constants but we can avoid nested classes and simply add all constants directly under TestData class and access them directly like `TestData.CONSTANT_NAME` WDYT?

File: skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java [hunk 33]
```diff
@@ -0,0 +1,134 @@
...
+            .contentType(MediaType.APPLICATION_JSON))
+            .andExpect(MockMvcResultMatchers.status().isCreated())
+            .andExpect(MockMvcResultMatchers.content().string(RegistrationActionResponses.CREATED.statusMessage))
+            .andDo((result) -> Assert.assertTrue(mockUserService.exists(userUnderTest)))
+            .andReturn();
+    }
+
+    @Test
+    public void register_withInvalidEmail_returnsBadRequest() throws Exception {
+        userUnderTest.getContacts().setEmail("not an email");
```
Comment (adawood) [inline skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java]: Please add the string "not an email" as a constant to TestConstants like the rest

**Review by adawood — CHANGES_REQUESTED** [review]

File: skypalwarmup/JSUserRegistrationService/src/test/java/application/TestData.java (hunk 32 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/test/java/application/TestData.java]: Done

File: skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java (hunk 33 above)
Comment (karim-elngr) [reply skypalwarmup/JSUserRegistrationService/src/test/java/application/controller/UserControllerTest.java]: Done

**Review by adawood — APPROVED** [review]

File: skypalwarmup/pom.xml (hunk 15 above)
Comment (motamira) [reply skypalwarmup/pom.xml:19]: Asked SRE team to provide access

**Review by motamira — APPROVED** [review]
